File: solutions/ocaml/105-course-schedule-ii.ml
(*
  Problem 105: Course Schedule II (LeetCode 210)
  Difficulty: Med
  Language: OCaml
*)
let find_order n prereqs =
  let module IS = Set.Make(Int) in
  let graph = List.fold_left (fun g (a, b) ->
    let cur = try List.assoc b g with Not_found -> [] in
    (b, a :: cur) :: List.filter (fun (k,_) -> k <> b) g
  ) [] prereqs in
  let nbrs v = try List.assoc v graph with Not_found -> [] in
  let rec dfs v (path, done, order, cyc) =
    if cyc || IS.mem v done then (path, done, order, cyc)
    else if IS.mem v path then (path, done, order, true)
    else
      let (p, d, o, c) = List.fold_left
        (fun s nb -> dfs nb s)
        (IS.add v path, done, order, false) (nbrs v) in
      (p, IS.add v d, o @ [v], c) in
  let (_, _, order, cyc) = List.init n Fun.id |> List.fold_left
    (fun s i -> dfs i s) (IS.empty, IS.empty, [], false) in
  if cyc then [] else order

File: solutions/scala/105-course-schedule-ii.scala
/**
 * Problem 105: Course Schedule II (LeetCode 210)
 * Difficulty: Med
 * Language: Scala
 */
def findOrder(n: Int, prereqs: Array[Array[Int]]): Array[Int] = {
  val graph = prereqs.foldLeft(Map.empty[Int, List[Int]]) {
    case (g, Array(a, b)) => g + (b -> (a :: g.getOrElse(b, Nil)))
  }
  case class S(path: Set[Int], done: Set[Int], order: List[Int], cycle: Boolean)
  def dfs(v: Int, s: S): S =
    if (s.cycle || s.done(v)) s
    else if (s.path(v)) s.copy(cycle = true)
    else {
      val s2 = graph.getOrElse(v, Nil).foldLeft(s.copy(path = s.path + v))(
        (st, nb) => dfs(nb, st))
      s2.copy(done = s2.done + v, order = s2.order :+ v)
    }
  val result = (0 until n).foldLeft(S(Set(), Set(), Nil, false))((s, i) => dfs(i, s))
  if (result.cycle) Array() else result.order.toArray
}
